src/machines.jl

```diff
@@ -78,39 +79,192 @@ constructor for retraining on new data using the saved model.
+        else
+            # The same machine can be associated with multiple nodes
+            if N.machine in keys(newmach_given_old)
+                m = newmach_given_old[N.machine]
+            else
+                m = serializable(filename, N.machine)
+                m.args = Tuple(newnode_given_old[s] for s in N.machine.args)
+                newmach_given_old[N.machine] = m
+            end
+            newnode_given_old[N] = N.operation(m, args...)
+        end
+        # Sort nodes according to: operation_node, report_node
+        if N in operation_nodes
+            newoperation_node_given_old[N] = newnode_given_old[N]
+        elseif N in report_nodes
+            newreport_node_given_old[N] = newnode_given_old[N]
+        end
     end
-    mach.state = 1
-    mach.fitresult = fitresult
-    mach.report = report
+
+    newoperation_nodes = Tuple(newoperation_node_given_old[N] for N in
+            operation_nodes)
+    newreport_nodes = Tuple(newreport_node_given_old[N] for N in
+            report_nodes)
+    report_tuple =
+        NamedTuple{keys(MLJBase._report_part(signature))}(newreport_nodes)
+    operation_tuple =
+        NamedTuple{keys(MLJBase._operation_part(signature))}(newoperation_nodes)
+
+    newsignature = if isempty(report_tuple)
+                        operation_tuple
+                    else
+                        merge(operation_tuple, (report=report_tuple,))
+                    end
+    
+
+    newfitresult = MLJBase.CompositeFitresult(newsignature)
+    setfield!(newfitresult, :report_additions, report_tuple)
+
+    return newfitresult
+
+end
+
+###############################################################################
+#####                  SERIALIZABLE AND RESTORE                           #####
+###############################################################################
+
+
+"""
+    serializable(filename, mach::Machine; kwargs...)
+
+Copies the state of the machine to make it serializable.
+    - Removes all data from caches dans data fields
+    - Makes all `fitresults` serializable
+"""
+function serializable(filename, mach::Machine; kwargs...)
+    copymach = machine(mach.model, mach.args..., cache=typeof(mach).parameters[2])
+
+    for fieldname in fieldnames(Machine)
+        if fieldname ∈ (:model, :args, :report)
+            continue
+        # Wipe data from cache
+        elseif fieldname == :cache 
+            wipe_cached_data!(copymach, mach)
+        # Wipe data from data
+        elseif fieldname ∈ (:data, :resampled_data)
+            setfield!(copymach, fieldname, ())
+        # Make fitresult ready for serialization
+        elseif fieldname == :fitresult
+            copymach.fitresult = save(_filename(filename), mach.model, getfield(mach, fieldname), kwargs...)
+        else
+            setfield!(copymach, fieldname, getfield(mach, fieldname))
+        end
+    end
+
+    setreport!(copymach, mach.report)
+    
```

Comment:
Minor detail: Maybe we want to set the value of `copymach.state=-1` to tag the machine as "serialisable" so we can throw an error in the new situation that user tries `predict(mach, ....)` when `mach` has been stripped of data (logic that will need adding at `MLJBase/src/operations.jl`).

Reply:
I suppose that depends on how faithful you want the `state` to be and what is the current logic behind it. Let's imagine a situation where data comes in every day and we update our machine with it as it comes and serialize it back to disk. I suppose we would like to record how many times we have updated our machine in the `state`? Would that be captured at the moment? Setting it ot `-1`, will necessarily erase this information.

Reply:
Well, currently MLJ does not support incremental learning, that is "update machine with new data" (eg, gradient descent models) - only update with new model parameters or new views of the same data. If you get new data, you need to rebind your model with all the data you have and retrain from scratch. There was some plan to support incremental learning but in that case we would probably introduce a new field to keep track of the number of "data injections". 

Presently the `state` variable is used to determine: (i) has the machine been trained (`state > 0`) and (ii) whether a machine upstream of a machine `mach` in a learning network has been trained since the last time `mach` was trained (via the `old_upstream_state` field, which is a snapshot of the `state` of every upstream machine). 

On the other hand, if you have a reliable way to detect if a machine has been serialised, we can go with that.

Reply:
The problem with setting the `state` to `-1` is that when restored, the machine cannot be used again to `predict` unless we set the state back again to 1 in `restore` which seems a bit counterproductive. Any thought?

Reply:
Ah. You are right. My first thought is to change the `predict` logic to allow `-1`.